README.md
# DeliveryCalculator

[![Latest Version on Packagist][ico-version]][link-packagist]
[![Software License][ico-license]](LICENSE.md)
[![Build Status][ico-travis]][link-travis]
[![Coverage Status][ico-scrutinizer]][link-scrutinizer]
[![Quality Score][ico-code-quality]][link-code-quality]
[![Total Downloads][ico-downloads]][link-downloads]

Calculate delivery time - only working days, skip holidays or weekends (witch can be turned off, but by default is on).

## Install

Via Composer

``` bash
$ composer require johnykvsky/deliverycalculator
```

## Usage

``` php
$dlvCalc = new johnykvsky\Utils\DeliveryCalculator();
$provider = new johnykvsky\Utils\PolishDeliveryProvider(); //We want to ship/deliver in Poland
$dlvCalc->setShippingProvider($provider); //from Poland
$dlvCalc->setDeliveryProvider($provider); //to Poland
$result = $dlvCalc->calculateDeliveryDate(14, '2017-10-20'); //Get delivery date, shipping on 2017-10-20, deliver in 14 working days
echo $result->format('Y-m-d'); //2017-11-10
```

DeliveryCalculator depends on providers, however, you can create one generic provider and use it for all calculations, modifying on the fly (getters/setters) timezones, holidays and delivery-free days (ie. when client paid for delivery on sunday).

Providers has following settings:
- timezone
- holidays
- nonWorkingDays
- region

Timezone is used for proper delivery time calculation, with ie. 8h differences delivery date might skip to next day. Holidays are dates of christmas, easter and all other "officially free from work" days. NonWorkingDays are days when usually we don't deliver, like sunday. Region can be optionally set and provided with additional holidays, specific for this particular country area.

Delivery and shipping location has separate providers, witch is important, since if we want to ship tomorrow (item was sold today), but tomorrow is sunday, next day is holiday, so shipping should skip them and count from first working day.


## Testing

``` bash
$ composer test
```

## Security

If you discover any security related issues, please email johnykvsky@protonmail.com instead of using the issue tracker.

## Credits

- [johnykvsky][link-author]

## License

The MIT License (MIT). Please see [License File](LICENSE.md) for more information.

[ico-version]: https://img.shields.io/packagist/v/johnykvsky/DeliveryCalculator.svg?style=flat-square
[ico-license]: https://img.shields.io/badge/license-MIT-brightgreen.svg?style=flat-square
[ico-travis]: https://img.shields.io/travis/johnykvsky/DeliveryCalculator/master.svg?style=flat-square
[ico-scrutinizer]: https://img.shields.io/scrutinizer/coverage/g/johnykvsky/DeliveryCalculator.svg?style=flat-square
[ico-code-quality]: https://img.shields.io/scrutinizer/g/johnykvsky/DeliveryCalculator.svg?style=flat-square
[ico-downloads]: https://img.shields.io/packagist/dt/johnykvsky/DeliveryCalculator.svg?style=flat-square

[link-packagist]: https://packagist.org/packages/johnykvsky/DeliveryCalculator
[link-travis]: https://travis-ci.org/johnykvsky/DeliveryCalculator
[link-scrutinizer]: https://scrutinizer-ci.com/g/johnykvsky/DeliveryCalculator/code-structure
[link-code-quality]: https://scrutinizer-ci.com/g/johnykvsky/DeliveryCalculator
[link-downloads]: https://packagist.org/packages/johnykvsky/DeliveryCalculator
[link-author]: https://github.com/johnykvsky


CHANGELOG.md
# Changelog

All notable changes to `DeliveryCalculator` will be documented in this file.

Updates should follow the [Keep a CHANGELOG](http://keepachangelog.com/) principles.

## 0.0.1 - 2017-12-25

### Added
- DeliveryCalculator initial release

### Deprecated
- Nothing

### Fixed
- Nothing

### Removed
- Nothing

### Security
- Nothing


tests/ProviderTest.php
<?php

use PHPUnit\Framework\TestCase;

class ProviderTest extends TestCase
{
    public $provider;
    public $holidays;

    protected function setUp()
    {
        $this->provider = new johnykvsky\Utils\PolishDeliveryProvider();
        $this->holidays = array(
            0 => '2017-01-01',
            1 => '2017-01-06',
            2 => '2017-04-16',
            3 => '2017-04-17',
            4 => '2017-04-18',
            5 => '2017-05-01',
            6 => '2017-05-03',
            7 => '2017-06-04',
            8 => '2017-06-15',
            9 => '2017-08-15',
            10 => '2017-11-01',
            11 => '2017-11-11',
            12 => '2017-12-25',
            13 => '2017-12-26'
        );
    }

    public function testSetTimezone()
    {
        $this->provider->setTimezone('Europe/Warsaw');
        $this->assertEquals('Europe/Warsaw', $this->provider->timezone);
    }

    public function testGetDateTimeZone()
    {
        $timezone =  new \DateTimeZone('Europe/Warsaw');
        $this->assertEquals($timezone, $this->provider->getDateTimeZone());
    }

    public function testGetNonWorkingWeekDays()
    {
        $this->assertEquals(array(0,6), $this->provider->getNonWorkingWeekDays());
    }

    public function testSetNonWorkingWeekDays()
    {
        $this->provider->setNonWorkingWeekDays(array(0,6));
        $this->assertEquals(array(0,6), $this->provider->getNonWorkingWeekDays());
    }

    public function testSetRegion()
    {
        $this->provider->setRegion('quacamole');
        $this->assertEquals('quacamole', $this->provider->region);
    }

    public function testGetHolidays()
    {
        $this->assertEquals($this->holidays, $this->provider->getHolidays('2017'));
    }

    public function testAddHoliday()
    {
        $reversed = array_reverse($this->holidays);
        $reversed[14] = '2017-08-01';
        $this->provider->addHoliday('2017-08-01');
        $this->assertEquals(array_reverse($reversed), $this->provider->getHolidays('2017'));
    }

    public function testAddRegionHolidays()
    {
        $this->assertEquals(null, $this->provider->addRegionHolidays('2017', '2017-04-16'));
    }
}


tests/DeliveryCalculatorTest.php
<?php

use PHPUnit\Framework\TestCase;

class DeliveryCalculatorTest extends TestCase
{
    public $dlvCalc;
    public $provider;

    protected function setUp()
    {
        $this->dlvCalc = new johnykvsky\Utils\DeliveryCalculator();
        $this->provider = new johnykvsky\Utils\PolishDeliveryProvider();
        $this->dlvCalc->setShippingProvider($this->provider);
        $this->dlvCalc->setDeliveryProvider($this->provider);
    }

    public function testSetShippingProvider()
    {
        $this->dlvCalc->setShippingProvider($this->provider);
        $this->assertEquals($this->provider, $this->dlvCalc->shippingProvider);
    }

    public function testSetDeliveryProvider()
    {
        $this->dlvCalc->setDeliveryProvider($this->provider);
        $this->assertEquals($this->provider, $this->dlvCalc->deliveryProvider);
    }

    public function testSetAdditionalNonWorkingDays()
    {
        $this->dlvCalc->setAdditionalNonWorkingDays(array(1));
        $this->assertEquals(array(1), $this->dlvCalc->additionalNonWorkingDays);
    }

    public function testGetAdditionalNonWorkingDays()
    {
        $this->dlvCalc->setAdditionalNonWorkingDays(array(1));
        $this->assertEquals(array(1), $this->dlvCalc->getAdditionalNonWorkingDays());
    }

    public function testCalculateDeliveryDate()
    {
        $result = $this->dlvCalc->calculateDeliveryDate(14, '2017-10-20')->format('Y-m-d');
        $this->assertEquals('2017-11-10', $result);
    }

    public function testYearChangeWithDifferentTimezoneDelivery()
    {
        $provider = new johnykvsky\Utils\PolishDeliveryProvider();
        $provider->timezone = 'Pacific/Chatham';
        $this->dlvCalc->setShippingProvider($provider);
        $result = $this->dlvCalc->calculateDeliveryDate(5, '2017-12-26')->format('Y-m-d');
        $this->assertEquals('2018-01-02', $result);
    }

    public function testBadTimezone()
    {
        $provider = new johnykvsky\Utils\PolishDeliveryProvider();
        $provider->timezone = 'Pacific/Chatham1';
        $this->expectException(\Exception::class);
        $this->dlvCalc->setShippingProvider($provider);
        $result = $this->dlvCalc->calculateDeliveryDate(4, '2017-12-26');
    }
}


src/DeliveryCalculator.php
<?php

namespace johnykvsky\Utils;

/**
 * Calculate delivery time, skip holidays and/or weekends
 */
class DeliveryCalculator
{
    /**
     * @var mixed $shippingProvider Provider for shipping timezone and holidays
     */
    public $shippingProvider;

    /**
     * @var mixed $deliveryProvider Provider for delivery timezone and holidays
     */
    public $deliveryProvider;

    /**
     * @var mixed $additionalNonWorkingDays Additional non working days (ie. we don't deliver on mondays)
     */
    public $additionalNonWorkingDays = [];

    /**
      * Set shipping provider
      *
      * @param mixed $shippingProvider Provider
      *
      * @return void
      */
    public function setShippingProvider($shippingProvider)
    {
        $this->shippingProvider = $shippingProvider;
    }

    /**
      * Set delivery provider
      *
      * @param mixed $deliveryProvider Provider
      *
      * @return void
      */
    public function setDeliveryProvider($deliveryProvider)
    {
        $this->deliveryProvider = $deliveryProvider;
    }

    /**
      * Set additional working days
      *
      * @param mixed $additionalNonWorkingDays Additional working days
      *
      * @return void
      */
    public function setAdditionalNonWorkingDays($additionalNonWorkingDays)
    {
        $this->additionalNonWorkingDays = $additionalNonWorkingDays;
    }

    /**
      * Get additional working days
      *
      * @return array
      */
    public function getAdditionalNonWorkingDays()
    {
        return $this->additionalNonWorkingDays;
    }

    /**
      * Calculates delivery date in X days from Y date
      *
      * @param integer $deliveryInDays Deliver package in X days
      * @param string $shippingDate Shipping start date, Y-m-d format
      *
      * @return \DateTime
      */
    public function calculateDeliveryDate($deliveryInDays, $shippingDate)
    {
        $deliverableDaysCounter = 0;
        $dayCounter = 1;

        try {
            $shippingTimezone = $this->shippingProvider->getDateTimeZone();
            $deliveryTimezone = $this->deliveryProvider->getDateTimeZone();
            $shippingDate = \DateTime::createFromFormat('Y-m-d', $shippingDate, $shippingTimezone);
            $initialShippingDate = \DateTimeImmutable::createFromMutable($shippingDate);
        } catch (\Exception $e) {
            throw $e;
        }

        $shippingHolidays = $this->shippingProvider->getHolidays($shippingDate->format('Y'));
        $deliveryHolidays = $this->deliveryProvider->getHolidays($shippingDate->format('Y'));

        while ($deliverableDaysCounter <= $deliveryInDays) {
            $deliveryDate = $shippingDate->modify('+1 day');

            if ($initialShippingDate->format('Y') != $deliveryDate->format('Y')) {
                $shippingHolidays = $this->shippingProvider->getHolidays($deliveryDate->format('Y'));
                $deliveryHolidays = $this->deliveryProvider->getHolidays($deliveryDate->format('Y'));
            }

            if ((!in_array($deliveryDate->format('Y-m-d'), $shippingHolidays)
                    or !in_array($deliveryDate->format('Y-m-d'), $deliveryHolidays))
                    and !in_array($deliveryDate->format('w'), $this->shippingProvider->getNonWorkingWeekDays())
                    and !in_array($deliveryDate->format('w'), $this->deliveryProvider->getNonWorkingWeekDays())
                    and !in_array($deliveryDate->format('w'), $this->getAdditionalNonWorkingDays())) { //no delivery in saturday 6, and sunday 0
                $deliverableDaysCounter++;
            }

            if ($deliverableDaysCounter == $deliveryInDays) {
                break;
            }
        }

        if (!empty($deliveryTimezone)) {
            $deliveryDate->setTimezone($deliveryTimezone);
        }
        return $deliveryDate;
    }
}


src/DeliveryProviderInterface.php
<?php

namespace johnykvsky\Utils;

/**
 * Interface for delivery providers
 */
interface DeliveryProviderInterface
{
    /**
      * Set timezone
      *
      * @param string $timezone Timezone name
      *
      * @return void
      */
    public function setTimezone($timezone);

    /**
      * Set region
      *
      * @param string $region Delivery region
      *
      * @return void
      */
    public function setRegion($region);

    /**
      * Add work-free day
      *
      * @param string $date Date in Y-m-d format
      *
      * @return void
      */
    public function addHoliday($date);

    /**
      * Get DateTimeZone with providers timezone
      *
      * @return \DateTimeZone
      */
    public function getDateTimeZone();

    /**
      * Get non working days
      *
      * @return array
      */
    public function getNonWorkingWeekDays();

    /**
      * {@inheritdoc}
      */
    public function setNonWorkingWeekDays($nonWorkingWeekDays);

    /**
      * Get dates of holidays for given year
      *
      * @param string $year Year for calculation
      *
      * @return void
      */
    public function getHolidays($year);

    /**
      * Add holidays specific for country region
      *
      * @param string $year Year
      * @param string $easter Easter day
      *
      * @return void
      */
    public function addRegionHolidays($year, $easter);
}


src/PolishDeliveryProvider.php
<?php

namespace johnykvsky\Utils;

/**
 * Calculate delivery time, only workdays
 */
class PolishDeliveryProvider implements \johnykvsky\Utils\DeliveryProviderInterface
{
    /**
     * @var string $timezone Timezone name
     */
    public $timezone = 'Europe/Warsaw';

    /**
     * @var array $holidays Holidays dates
     */
    public $holidays = [];

    /**
     * @var array $nonWorkingDays Non working days (weekends)
     */
    public $nonWorkingDays = array('0','6'); //no delivery on saturday 6, and sunday 0

    /**
     * @var string $region Region to deliver
     */
    public $region;

    /**
      * {@inheritdoc}
      */
    public function setTimezone($timezone)
    {
        $this->timezone = $timezone;
    }

    /**
      * {@inheritdoc}
      */
    public function setRegion($region)
    {
        $this->region = $region;
    }

    /**
      * {@inheritdoc}
      */
    public function addHoliday($date)
    {
        $this->holidays[] = $date;
    }

    /**
      * {@inheritdoc}
      */
    public function getDateTimeZone()
    {
        return new \DateTimeZone($this->timezone);
    }

    /**
      * {@inheritdoc}
      */
    public function getNonWorkingWeekDays()
    {
        return $this->nonWorkingDays;
    }

    /**
      * {@inheritdoc}
      */
    public function setNonWorkingWeekDays($nonWorkingWeekDays)
    {
        $this->nonWorkingDays = $nonWorkingWeekDays;
    }

    /**
      * {@inheritdoc}
      */
    public function getHolidays($year)
    {
        $timezone = $this->getDateTimeZone();
        $date = \DateTime::createFromFormat('Y-m-d', $year.'-01-01', $timezone);
        $year = $date->format('Y');

        $easter = easter_date($year);

        $this->addHoliday($year.'-01-01'); //Nowy Rok, Świętej Bożej Rodzicielki
        $this->addHoliday($year.'-01-06'); //Trzech Króli (Objawienie Pańskie)
        $this->addHoliday(date('Y-m-d', $easter)); //Wielkanocna sobota, ruchome święto
        $this->addHoliday(date('Y-m-d', strtotime('+1 day', $easter))); //Wielkanocna niedziela, ruchome święto
        $this->addHoliday(date('Y-m-d', strtotime('+2 day', $easter))); //Poniedziałek Wielkanocny
        $this->addHoliday($year.'-05-01'); //Święto Pracy
        $this->addHoliday($year.'-05-03'); //Święto Konstytucji 3 Maja
        $this->addHoliday(date('Y-m-d', strtotime('+49 days', $easter))); //Zesłanie Ducha Świętego (Zielone Świątki), 49 dni po niedzieli wielkanocnej
        $this->addHoliday(date('Y-m-d', strtotime('+60 days', $easter))); //Boże Ciało, 60 dni po niedzieli wielkanocnej
        $this->addHoliday($year.'-08-15'); //Święto Wojska Polskiego, Wniebowzięcie Najświętszej Maryi Panny
        $this->addHoliday($year.'-11-01'); //Wszystkich Świętych
        $this->addHoliday($year.'-11-11'); //Święto Niepodległości
        $this->addHoliday($year.'-12-25'); //Boże Narodzenie (pierwszy dzień)
        $this->addHoliday($year.'-12-26'); //Boże Narodzenie (pierwszy dzień)

        $this->addRegionHolidays($year, $easter);

        return $this->holidays;
    }

    /**
      * {@inheritdoc}
      */
    public function addRegionHolidays($year, $easter)
    {
        return;
    }
}
